Sample project to show the grapher capabilities (#6)
FundingCircle/topology-grapher: README.md

[![Clojars Project](https://img.shields.io/clojars/v/fundingcircle/topology-grapher.svg)](https://clojars.org/fundingcircle/topology-grapher)
[![CircleCI](https://circleci.com/gh/FundingCircle/topology-grapher/tree/master.svg?style=svg)](https://circleci.com/gh/FundingCircle/topology-grapher/tree/master)
[![codecov](https://codecov.io/gh/FundingCircle/topology-grapher/branch/master/graph/badge.svg)](https://codecov.io/gh/FundingCircle/topology-grapher)

# Topology Grapher

This library provides a means to build a directed graph (data) from a kafka streams topology.

> # How to integrate your application

Reviewer: ```suggestion
## How to integrate into your application
```

I think it‘s best for Markdown docs to have only a single h1 header.

If you want to play around with this library you can check the [sample project](./sample_project/README.md).

There are a few simple steps to follow to integrate your kafka streaming app.

> ## 1. implement a command to call the describe-topologies

Reviewer: ```suggestion
### 1. Create a namespace or function that invokes `generate-zip`
```

```
(ns your-namespace
  (:require [topology-grapher.describe :refer [generate-zip]]))

(def meta-data {:domain "domain"
    :subdomain "subdomain"
    :application "application-name"})

(def topologies
  [{:application-name "app-2"
    :topology topology-1}

   {:application-name "app-1"
    :topology topology-2}])

(generate-zip topologies meta-data)

```

The `generate-zip` function is what takes care of transforming
topologies objects into EDN files that are written out to disk.

The important thing is that you need to be have a function that
returns a topology object *without* having to actually start Kafka.

This is normally fine, but if you are using something like the Stuart

> Sierra components library for example you will need to refactor the

Reviewer: ```suggestion
Sierra [component](https://github.com/stuartsierra/component) for example you will need to refactor the
```

initialisation to avoid needing the full system started.

> ## 2. add a call to that command in CircleCI

Reviewer: ```suggestion
### 2. Create a CI job that invokes that namespace or function
```

> This is an example of configuration you need to add to CircleCI to generate and publish graph information.

Reviewer: ```suggestion
Here’s an example [CircleCI](https://circleci.com/) job that generates and publishes the graph data:
```

> ```

Reviewer: ```suggestion
```yaml
```

  publish_topologies_graphs:
    docker:

>       - image: *lein

Reviewer: This `*lein` value seems odd to me. What does it mean? Is it a valid value or a placeholder of some kind?

Author: it's a variable, but yeah confusing, I'll use a real image example instead

    steps:
      - checkout
      - run: sudo apt update && sudo apt install awscli
      - restore_cache:
          key: your-project-{{ checksum "project.clj" }}

      # the precise command depends on how you integrated
      - run: lein run describe-topologies
      # not the real bucket yet since we are still waiting for one
      # - run: aws s3 cp --recursive /tmp/graphs/$your-project-name s3://$your-bucket
```

> This particular CircleCI step can run in parallel with any other job, and should not be required as dependency.

Reviewer: ```suggestion
This CircleCI job can run in parallel with any other job.
```

I don’t think we need to provide quite so much detailed guidance on integrating with CircleCI.

And you are done, if you integrated in a way that all the topologies will always be published automatically

> there is 0 maintenance effort needed from now on.

Reviewer: This paragraph seems unnecessary.

> # Graph implementation

Reviewer: ```suggestion
## Graph implementation
```

I think it‘s best for Markdown docs to have only a single h1 header.

The graph generation has been tested on Kafka 1.0+ streams apps, both using the
higher level DSL and lower level streams APIs.

> ## Internal graph format

Reviewer: ```suggestion
### Internal graph format
```

I think it‘s best for Markdown docs to have only a single h1 header.

* `lein test`
* `lein install` (to install a local snapshot for any local integration testing)

> # Release

Reviewer: ```suggestion
## Release
```

I think it‘s best for Markdown docs to have only a single h1 header.

- run `lein deploy` to do the release

> # Contributing

Reviewer: ```suggestion
## Contributing
```

I think it‘s best for Markdown docs to have only a single h1 header.
